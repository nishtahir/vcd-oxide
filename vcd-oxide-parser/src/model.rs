use core::cell::RefCell;
use std::{
    collections::BTreeMap,
    rc::{Rc, Weak},
};

type VcdScopeNode = Rc<RefCell<ValueChangeDumpScope>>;

#[derive(Debug, Default)]
pub struct ValueChangeDump {
    pub date: String,
    pub version: String,
    pub timescale: String,
    // TODO: Validate this but there's only one root scope
    // This scope can have multiple child scopes
    pub root_scope: VcdScopeNode,
    pub wave_map: BTreeMap<String, ValueChangeDumpWave>,
}

#[derive(Debug, Default)]
pub struct ValueChangeDumpScope {
    pub name: String,
    pub kind: String,
    pub scopes: Vec<Rc<RefCell<ValueChangeDumpScope>>>,
    pub parent: Option<Weak<RefCell<ValueChangeDumpScope>>>,
    pub signals: Vec<ValueChangeDumpSignal>,
}

#[derive(Debug, Default)]
pub struct ValueChangeDumpSignal {
    pub kind: String,
    pub identifier: String,
    pub reference: String,
    pub size: usize,
}

#[derive(Debug, Default)]
pub struct ValueChangeDumpWave {
    pub value_changes: Vec<ValueChange>,
}

#[derive(Debug, Default)]
pub struct ValueChange {
    pub time: usize,
    pub value: String,
}
