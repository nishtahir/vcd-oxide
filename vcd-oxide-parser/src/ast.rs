#[derive(Debug)]
pub struct ValueChangeDumpDefinition {
    pub declaration_commands: Vec<DeclarationCommand>,
    pub simulation_commands: Vec<SimulationCommand>,
}

#[derive(Debug)]
pub enum DeclarationCommand {
    Comment(GenericComment),
    Date(DeclarationDate),
    EndDefinitions,
    Scope(DeclarationScope),
    Timescale(DeclarationTimescale),
    Upscope,
    Var(DeclarationVar),
    Version(DeclarationVersion),
}

#[derive(Debug)]
pub struct DeclarationDate {
    pub value: String,
}

#[derive(Debug)]
pub struct DeclarationVersion {
    pub value: String,
}

#[derive(Debug)]
pub struct DeclarationTimescale {
    pub time_number: usize,
    pub time_unit: String,
}

#[derive(Debug)]
pub struct DeclarationScope {
    pub scope_type: String,
    pub scope_identifier: String,
}

#[derive(Debug)]
pub struct DeclarationVar {
    pub var_type: String,
    pub size: usize,
    pub identifier_code: String,
    pub reference: String,
}

#[derive(Debug)]
pub enum SimulationCommand {
    KeywordCommand(SimulationKeywordCommand),
    Comment(GenericComment),
    SimulationTime,
    ValueChange(SimulationValueChange),
}

#[derive(Debug)]
pub struct SimulationKeywordCommand {
    pub ty: SimulationType,
    pub value_changes: Vec<SimulationValueChange>,
}

#[derive(Debug)]
pub enum SimulationType {
    DumpAll,
    DumpOff,
    DumpOn,
    DumpVars,
}

#[derive(Debug)]
pub enum SimulationValueChange {
    Scalar(ScalarValueChange),
    Vector(VectorValueChange),
}

#[derive(Debug)]
pub struct ScalarValueChange {
    pub value: String,
    pub identifier_code: String,
}

#[derive(Debug)]
pub enum VectorValueChange {
    Binary(BinaryVectorValueChange),
    Real(RealVectorValueChange),
}

#[derive(Debug)]
pub struct BinaryVectorValueChange {
    pub value: String,
    pub identifier_code: String,
}

#[derive(Debug)]
pub struct RealVectorValueChange {
    pub value: String,
    pub identifier_code: String,
}

#[derive(Debug)]
pub struct GenericComment {
    pub value: String,
}