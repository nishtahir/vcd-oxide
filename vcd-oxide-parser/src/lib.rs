mod ast;
mod model;

extern crate pest;
extern crate pest_derive;

use crate::ast::*;
use crate::model::*;
use pest::{iterators::Pair, Parser};
use pest_derive::Parser;
use std::collections::HashMap;
use std::{cell::RefCell, rc::Rc};

#[derive(Parser, Debug)]
#[grammar = "grammar/vcd.pest"]
pub struct ValueChangeDumpParser;

fn visit_value_change_dump_definitions(rule: Pair<Rule>) -> ValueChangeDumpDefinition {
    let mut declaration_commands = vec![];
    let mut simulation_commands = vec![];

    for item in rule.into_inner() {
        match item.as_rule() {
            Rule::declaration_command => {
                let declaration_command = visit_declaration_command(item);
                declaration_commands.push(declaration_command);
            }
            Rule::simulation_command => {
                let simulation_command = visit_simulation_command(item);
                simulation_commands.push(simulation_command);
            }
            _ => unreachable!("{:#?}", item),
        };
    }

    ValueChangeDumpDefinition {
        declaration_commands,
        simulation_commands,
    }
}

fn visit_simulation_command(rule: Pair<Rule>) -> SimulationCommand {
    let inner = rule.into_inner().next().unwrap();
    match inner.as_rule() {
        Rule::vcd_simulation_dumpall => visit_vcd_simulation_dumpall(inner),
        Rule::vcd_simulation_dumpoff => visit_vcd_simulation_dumpoff(inner),
        Rule::vcd_simulation_dumpon => visit_vcd_simulation_dumpon(inner),
        Rule::vcd_simulation_dumpvars => visit_vcd_simulation_dumpvars(inner),
        Rule::simulation_keyword_comment => visit_simulation_keyword_comment(inner),
        Rule::simulation_time => visit_simulation_time(inner),
        Rule::value_change => SimulationCommand::ValueChange(visit_value_change(inner)),
        _ => unreachable!("{:#?}", inner),
    }
}

fn visit_vcd_simulation_dumpall(rule: Pair<Rule>) -> SimulationCommand {
    SimulationCommand::KeywordCommand(SimulationKeywordCommand {
        ty: SimulationType::DumpAll,
        value_changes: visit_value_change_list(rule),
    })
}

fn visit_vcd_simulation_dumpoff(rule: Pair<Rule>) -> SimulationCommand {
    SimulationCommand::KeywordCommand(SimulationKeywordCommand {
        ty: SimulationType::DumpOff,
        value_changes: visit_value_change_list(rule),
    })
}

fn visit_vcd_simulation_dumpon(rule: Pair<Rule>) -> SimulationCommand {
    SimulationCommand::KeywordCommand(SimulationKeywordCommand {
        ty: SimulationType::DumpOn,
        value_changes: visit_value_change_list(rule),
    })
}

fn visit_vcd_simulation_dumpvars(rule: Pair<Rule>) -> SimulationCommand {
    SimulationCommand::KeywordCommand(SimulationKeywordCommand {
        ty: SimulationType::DumpVars,
        value_changes: visit_value_change_list(rule),
    })
}

fn visit_simulation_keyword_comment(rule: Pair<Rule>) -> SimulationCommand {
    SimulationCommand::Comment(GenericComment {
        value: rule.as_str().to_owned(),
    })
}

fn visit_simulation_time(rule: Pair<Rule>) -> SimulationCommand {
    let value = rule
        .into_inner()
        .next()
        .unwrap()
        .as_str()
        .parse::<usize>()
        .unwrap();
    SimulationCommand::SimulationTime(SimulationTime { value })
}

fn visit_value_change_list(rule: Pair<Rule>) -> Vec<SimulationValueChange> {
    let mut inner = rule.into_inner();
    let mut value_changes = vec![];
    for i in inner {
        value_changes.push(visit_value_change(i));
    }
    value_changes
}

fn visit_value_change(rule: Pair<Rule>) -> SimulationValueChange {
    let inner = rule.into_inner().next().unwrap();
    match inner.as_rule() {
        Rule::scalar_value_change => visit_scalar_value_change(inner),
        Rule::vector_value_change => visit_vector_value_change(inner),
        _ => unreachable!("{:#?}", inner),
    }
}

fn visit_scalar_value_change(rule: Pair<Rule>) -> SimulationValueChange {
    let mut inner = rule.into_inner();
    let value = inner.next().unwrap();
    let identifier_code = inner.next().unwrap();
    SimulationValueChange::Scalar(ScalarValueChange {
        value: value.as_str().to_owned(),
        identifier_code: identifier_code.as_str().to_owned(),
    })
}

fn visit_vector_value_change(rule: Pair<Rule>) -> SimulationValueChange {
    let mut inner = rule.into_inner();
    let value = inner.next().unwrap();
    let identifier_code = inner.next().unwrap();
    SimulationValueChange::Vector(match value.as_rule() {
        Rule::real_value => VectorValueChange::Real(RealVectorValueChange {
            value: value.as_str().to_owned(),
            identifier_code: identifier_code.as_str().to_owned(),
        }),
        Rule::binary_value => VectorValueChange::Binary(BinaryVectorValueChange {
            value: value.as_str().to_owned(),
            identifier_code: identifier_code.as_str().to_owned(),
        }),
        _ => unreachable!("{:#?}", inner),
    })
}

fn visit_declaration_command(rule: Pair<Rule>) -> DeclarationCommand {
    let inner = rule.into_inner().next().unwrap();
    match inner.as_rule() {
        Rule::vcd_declaration_vars => DeclarationCommand::Var(visit_vcd_declaration_vars(inner)),
        Rule::vcd_declaration_comment => {
            DeclarationCommand::Comment(visit_vcd_declaration_comment(inner))
        }
        Rule::vcd_declaration_date => DeclarationCommand::Date(visit_vcd_declaration_date(inner)),
        Rule::vcd_declaration_enddefinitions => DeclarationCommand::EndDefinitions,
        Rule::vcd_declaration_scope => {
            DeclarationCommand::Scope(visit_vcd_declaration_scope(inner))
        }
        Rule::vcd_declaration_timescale => {
            DeclarationCommand::Timescale(visit_vcd_declaration_timescale(inner))
        }
        Rule::vcd_declaration_upscope => DeclarationCommand::Upscope,
        Rule::vcd_declaration_version => {
            DeclarationCommand::Version(visit_vcd_declaration_version(inner))
        }
        _ => unreachable!("{:#?}", inner),
    }
}

fn visit_vcd_declaration_date(rule: Pair<Rule>) -> DeclarationDate {
    let inner = rule.into_inner().next().unwrap();
    DeclarationDate {
        value: inner.as_str().to_owned(),
    }
}

fn visit_vcd_declaration_version(rule: Pair<Rule>) -> DeclarationVersion {
    let inner = rule.into_inner().next().unwrap();
    DeclarationVersion {
        value: inner.as_str().to_owned(),
    }
}

fn visit_vcd_declaration_timescale(rule: Pair<Rule>) -> DeclarationTimescale {
    let mut inner = rule.into_inner();
    let time_number = inner.next().unwrap().as_str().parse::<usize>().unwrap();
    let time_unit = inner.next().unwrap().as_str().to_owned();

    DeclarationTimescale {
        time_number,
        time_unit,
    }
}

fn visit_vcd_declaration_scope(rule: Pair<Rule>) -> DeclarationScope {
    let mut inner = rule.into_inner();
    let scope_type = inner.next().unwrap().as_str().to_owned();
    let scope_identifier = inner.next().unwrap().as_str().to_owned();

    DeclarationScope {
        scope_type,
        scope_identifier,
    }
}

fn visit_vcd_declaration_vars(rule: Pair<Rule>) -> DeclarationVar {
    let mut inner = rule.into_inner();
    let var_type = inner.next().unwrap().as_str().to_owned();
    let size = inner.next().unwrap().as_str().parse::<usize>().unwrap();
    let identifier_code = inner.next().unwrap().as_str().to_owned();
    let reference = inner.next().unwrap().as_str().to_owned();

    DeclarationVar {
        var_type,
        size,
        identifier_code,
        reference,
    }
}

fn visit_vcd_declaration_comment(rule: Pair<Rule>) -> GenericComment {
    let inner = rule.into_inner().next().unwrap();
    GenericComment {
        value: inner.as_str().to_owned(),
    }
}

fn visit_command_text(rule: Option<Pair<Rule>>) -> String {
    return rule.map_or("", |r| r.as_str()).to_owned();
}

fn parse(input: &str) -> ValueChangeDumpDefinition {
    let mut root = ValueChangeDumpParser::parse(Rule::file, input).unwrap();
    let inner = root.next().unwrap();
    match inner.as_rule() {
        Rule::value_change_dump_definitions => visit_value_change_dump_definitions(inner),
        _ => unreachable!("{:#?}", inner),
    }
}

impl ValueChangeDump {
    pub fn fromDefinition(definition: ValueChangeDumpDefinition) -> Self {
        let mut dump = ValueChangeDump::default();
        let mut active_scope = dump.root_scope.clone();
        for declaration in definition.declaration_commands {
            match declaration {
                DeclarationCommand::Comment(_) => {
                    // Ignore
                }
                DeclarationCommand::EndDefinitions => {
                    // Ignore
                }
                DeclarationCommand::Date(date) => dump.date = date.value,
                DeclarationCommand::Timescale(timescale) => {
                    dump.timescale = format!("{}{}", timescale.time_number, timescale.time_unit)
                }
                DeclarationCommand::Scope(scope) => {
                    let scope = Rc::<RefCell<ValueChangeDumpScope>>::new(RefCell::new(
                        ValueChangeDumpScope {
                            name: scope.scope_identifier,
                            parent: Some(Rc::<RefCell<ValueChangeDumpScope>>::downgrade(
                                &active_scope,
                            )),
                            ..Default::default()
                        },
                    ));
                    active_scope.borrow_mut().scopes.push(scope.clone());
                    active_scope = scope;
                }
                DeclarationCommand::Upscope => {
                    active_scope = active_scope
                        .clone()
                        .borrow()
                        .parent
                        .as_ref()
                        .unwrap()
                        .upgrade()
                        .unwrap();
                }
                DeclarationCommand::Var(var) => {
                    let signal_id = var.identifier_code;
                    active_scope
                        .borrow_mut()
                        .signals
                        .push(ValueChangeDumpSignal {
                            kind: var.var_type.to_owned(),
                            identifier: signal_id.to_owned(),
                            reference: var.reference.to_owned(),
                            size: var.size,
                            ..Default::default()
                        });

                    dump.wave_map
                        .insert(signal_id.to_owned(), ValueChangeDumpWave::default());
                }
                DeclarationCommand::Version(version) => dump.version = version.value,
            }
        }

        let mut simulation_time = 0;
        for sim in definition.simulation_commands {
            match sim {
                SimulationCommand::KeywordCommand(command) => {
                    for elem in &command.value_changes {
                        dump.wave_map
                            .get_mut(&elem.identifier_code())
                            .unwrap()
                            .value_changes
                            .push(ValueChange {
                                value: elem.value(),
                                time: simulation_time,
                            });
                    }
                }
                SimulationCommand::Comment(_) => {
                    // Ignore
                }
                SimulationCommand::SimulationTime(time) => simulation_time = time.value,
                SimulationCommand::ValueChange(value_change) => {
                    dump.wave_map
                        .get_mut(&value_change.identifier_code())
                        .unwrap()
                        .value_changes
                        .push(ValueChange {
                            value: value_change.value(),
                            time: simulation_time,
                        });
                }
            }
        }

        dump
    }

    fn create_value_change(time: usize, svc: SimulationValueChange) -> ValueChange {
        match svc {
            SimulationValueChange::Scalar(s) => ValueChange {
                time,
                value: s.value,
            },
            SimulationValueChange::Vector(v) => match v {
                VectorValueChange::Binary(b) => ValueChange {
                    time,
                    value: b.value,
                },
                VectorValueChange::Real(r) => ValueChange {
                    time,
                    value: r.value,
                },
            },
        }
    }
}

impl SimulationValueChange {
    fn identifier_code(&self) -> String {
        let s = match self {
            SimulationValueChange::Scalar(s) => &s.identifier_code,
            SimulationValueChange::Vector(v) => match v {
                VectorValueChange::Binary(b) => &b.identifier_code,
                VectorValueChange::Real(r) => &r.identifier_code,
            },
        };
        s.to_owned()
    }

    fn value(&self) -> String {
        let s = match self {
            SimulationValueChange::Scalar(s) => &s.value,
            SimulationValueChange::Vector(v) => match v {
                VectorValueChange::Binary(b) => &b.value,
                VectorValueChange::Real(r) => &r.value,
            },
        };
        s.to_owned()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_vcd_declaration_command() {
        let declerations = include_str!("../test/declaration_command.vcd.test");
        let ast = parse(declerations);
        assert_debug_snapshot!(ast)
    }

    #[test]
    fn test_vcd_file() {
        let declerations = include_str!("../test/UartRxTest.vcd");
        let ast = parse(declerations);
        assert_debug_snapshot!(ast)
    }

    #[test]
    #[ignore = "Runs for > 350s. Requires parser rewrite to be reasonable"]
    fn test_large_vcd_file() {
        let declerations = include_str!("../test/NextCoreTest.vcd");
        let ast = parse(declerations);
        assert_debug_snapshot!(ast)
    }

    #[test]
    fn test_model_against_uart() {
        let declerations = include_str!("../test/UartRxTest.vcd");
        let ast = parse(declerations);
        let model = ValueChangeDump::fromDefinition(ast);
        assert_debug_snapshot!(model)
    }
}
