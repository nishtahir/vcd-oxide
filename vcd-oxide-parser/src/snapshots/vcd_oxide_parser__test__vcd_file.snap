---
source: vcd-oxide-parser/src/lib.rs
assertion_line: 394
expression: ast
---
ValueChangeDumpDefinition {
    declaration_commands: [
        Date(
            DeclarationDate {
                value: "Fri Jan 27 10:13:28 2023",
            },
        ),
        Version(
            DeclarationVersion {
                value: "Icarus Verilog",
            },
        ),
        Timescale(
            DeclarationTimescale {
                time_number: 1,
                time_unit: "s",
            },
        ),
        Scope(
            DeclarationScope {
                scope_type: "module",
                scope_identifier: "UartRxTest",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 8,
                identifier_code: "!",
                reference: "data [7:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "\"",
                reference: "clk ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "#",
                reference: "rst ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "$",
                reference: "uart_rx ",
            },
        ),
        Scope(
            DeclarationScope {
                scope_type: "module",
                scope_identifier: "uart_rx_uut",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 1,
                identifier_code: "\"",
                reference: "clk ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 1,
                identifier_code: "%",
                reference: "read ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 1,
                identifier_code: "#",
                reference: "rst ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 1,
                identifier_code: "$",
                reference: "uart_rx ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 8,
                identifier_code: "&",
                reference: "data [7:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "wire",
                size: 1,
                identifier_code: "'",
                reference: "busy ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 32,
                identifier_code: "(",
                reference: "BAUD ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 32,
                identifier_code: ")",
                reference: "CLK_FREQUENCY_HZ ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 64,
                identifier_code: "*",
                reference: "DIV_COUNTER ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 32,
                identifier_code: "+",
                reference: "DIV_PER_BIT ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 32,
                identifier_code: ",",
                reference: "DIV_SAMPLE ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 64,
                identifier_code: "-",
                reference: "MAX_BAUD_COUNTER ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "parameter",
                size: 32,
                identifier_code: ".",
                reference: "MID_POINT ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 2,
                identifier_code: "/",
                reference: "baud_counter [1:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 4,
                identifier_code: "0",
                reference: "bit_counter [3:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "1",
                reference: "clear_bit ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "2",
                reference: "clear_sample ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 10,
                identifier_code: "3",
                reference: "data_register [9:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "4",
                reference: "inc_bit ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "5",
                reference: "inc_sample ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 2,
                identifier_code: "6",
                reference: "next_state [1:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 2,
                identifier_code: "7",
                reference: "sample_counter [1:0]",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 1,
                identifier_code: "8",
                reference: "shift ",
            },
        ),
        Var(
            DeclarationVar {
                var_type: "reg",
                size: 2,
                identifier_code: "9",
                reference: "state [1:0]",
            },
        ),
        Upscope,
        Upscope,
        EndDefinitions,
        Comment(
            GenericComment {
                value: "Show the parameter values.",
            },
        ),
    ],
    simulation_commands: [
        KeywordCommand(
            SimulationKeywordCommand {
                ty: DumpAll,
                value_changes: [
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b10",
                                identifier_code: ".",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b10",
                                identifier_code: "-",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b100",
                                identifier_code: ",",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b1010",
                                identifier_code: "+",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b10",
                                identifier_code: "*",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b1100100",
                                identifier_code: ")",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b1010",
                                identifier_code: "(",
                            },
                        ),
                    ),
                ],
            },
        ),
        SimulationTime(
            SimulationTime {
                value: 0,
            },
        ),
        KeywordCommand(
            SimulationKeywordCommand {
                ty: DumpVars,
                value_changes: [
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b10",
                                identifier_code: "9",
                            },
                        ),
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "x",
                            identifier_code: "8",
                        },
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "7",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b10",
                                identifier_code: "6",
                            },
                        ),
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "x",
                            identifier_code: "5",
                        },
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "x",
                            identifier_code: "4",
                        },
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "3",
                            },
                        ),
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "x",
                            identifier_code: "2",
                        },
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "x",
                            identifier_code: "1",
                        },
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "0",
                            },
                        ),
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "/",
                            },
                        ),
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "0",
                            identifier_code: "'",
                        },
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "&",
                            },
                        ),
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "z",
                            identifier_code: "%",
                        },
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "1",
                            identifier_code: "$",
                        },
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "1",
                            identifier_code: "#",
                        },
                    ),
                    Scalar(
                        ScalarValueChange {
                            value: "0",
                            identifier_code: "\"",
                        },
                    ),
                    Vector(
                        Binary(
                            BinaryVectorValueChange {
                                value: "b0",
                                identifier_code: "!",
                            },
                        ),
                    ),
                ],
            },
        ),
        SimulationTime(
            SimulationTime {
                value: 1,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "4",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "1",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "5",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "2",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "8",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "1",
                    identifier_code: "\"",
                },
            ),
        ),
        SimulationTime(
            SimulationTime {
                value: 2,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "\"",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "$",
                },
            ),
        ),
        SimulationTime(
            SimulationTime {
                value: 3,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "1",
                    identifier_code: "\"",
                },
            ),
        ),
        SimulationTime(
            SimulationTime {
                value: 4,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "\"",
                },
            ),
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "1",
                    identifier_code: "$",
                },
            ),
        ),
        SimulationTime(
            SimulationTime {
                value: 5,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "1",
                    identifier_code: "\"",
                },
            ),
        ),
        SimulationTime(
            SimulationTime {
                value: 6,
            },
        ),
        ValueChange(
            Scalar(
                ScalarValueChange {
                    value: "0",
                    identifier_code: "\"",
                },
            ),
        ),
    ],
}
